proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1189 @@
+  foo :: Show a => Code Q (a -> String)
+  foo = [|| show ||]
+
+However, there is actually a stage error introduced by this program as the
+evidence for ``Show a`` is bound earlier than it is used.
+The prototype correctly reports the following error::
+
+  TTH.hs:8:11: error: [GHC-28914]
+      • Stage error: ‘show’ is bound at stage {0} but used at stage 1
+        From imports {imported from ‘Prelude’ at TTH.hs:3:8-11}
+      • In the Template Haskell typed quotation [|| show ||]
+    |
+  8 | foo = [|| show ||]
+    |
+
+
+The language of constraints is not yet expressive enough to communicate that we
+want the ``Show a`` evidence to be available at a later stage. Fixing this problem
+will require
+significant additional effort, and there are other known issues with TTH (see
+`Staging with Class: a Specification for Typed Template Haskell
+<https://dl.acm.org/doi/abs/10.1145/3498723>`_). We propose that an initial
+implementation of ``NoImplicitStagePersistence`` may support untyped TH but not
+TTH (i.e. the compiler may reject programs using TTH under
+``NoImplicitStagePersistence``).  In the long term, we believe that implementing
+Staging with Class is desirable and consistent with the direction of travel
+established by this proposal, but the full details of Staging with Class are out
+of scope.
+
+
+Implicit lifting and deriving ``Lift`` instances
+################################################
+.. _lift_instances:
+
+``Lift`` instances are used to provide serialisation-based cross-stage
+persistence.  For example, a typical ``Lift`` instance looks like::
+
+    data MInt = Some Int | None
+
+    instance Lift MInt where
+        lift :: MInt -> Q Exp
+        lift None     = [| None |]
+        lift (Some x) = [| Some $(lift x) |]
+
+The presence of this instance means the following declaration will be accepted::
+
+  foo :: MInt -> Q Exp
+  foo x = [| x |]  -- implicitly becomes [| $(lift x) |]
+
+Defining a ``Lift`` instance requires the datatype constructors to be available
+both at compile-time and runtime, so defining ``Lift`` within the same module as
+the datatype itself requires path-based cross-stage persistence.  Operationally,
+``None`` and ``Some`` are needed both at compile-time *and*  runtime since they
+are both matched on at compile time, and also persisted to be spliced in the
+future into a program that can make use of them at runtime. As a result, it
+isn't possible to define or derive a (non-orphan) ``Lift`` instance under
+``NoImplicitStagePersistence``.
+
+An orphan ``Lift`` instance can be defined thus::
+
+  module M where
+    data MInt = Some Int | None
+
+  module N where
+    import M
+    import quote M
+
+    instance Lift MInt where
+        lift :: MInt -> Q Exp
+        lift None     = [| None |]
+        lift (Some x) = [| Some $(lift x) |]
+
+This isn't technically problematic, rather it is just a result of what ``Lift``
+means. However, it means some users may need to modify their use of ``Lift``
```

Comment:
Could you be more explicit about the consequences of `Lift` needing `NoImplicitStagePersistence`? I think it's worth repeating that `Lift` only requires `TemplateHaskellQuotes`, not full `TemplateHaskell`. This means there is no difference to modules importing `Lift` modules at stage *R* (correct?); they won't need it at stage *C*. However, modules that actually want to use `lift` at compile-time *C* (with a splice import) will also need to compile the `Lift` module for stage *R*. That's not too bad!

That would allow you to get rid of a lot of the "countermeasures" below ("as a general rule ..."), because users can infer by themselves whether it's worth separating out a data types module.